src/core.ts

```diff
@@ -636,53 +667,43 @@ export class NodePart implements Part {
     // of TemplateResults that will be commonly returned from expressions like:
     // array.map((i) => html`${i}`), by reusing existing TemplateInstances.
 
-    // If _previousValue is an array, then the previous render was of an
-    // iterable and _previousValue will contain the NodeParts from the previous
-    // render. If _previousValue is not an array, clear this part and make a new
+    // If _value is an array, then the previous render was of an
+    // iterable and _value will contain the NodeParts from the previous
+    // render. If _value is not an array, clear this part and make a new
     // array for NodeParts.
     if (!Array.isArray(this._value)) {
-      this.clear();
       this._value = [];
+      this.clear();
     }
 
     // Lets us keep track of how many items we stamped so we can clear leftover
     // items from a previous render
-    const itemParts = this._value as any[];
+    const itemParts = this._value as NodePart[];
     let partIndex = 0;
+    let itemPart: NodePart|undefined;
 
     for (const item of value) {
       // Try to reuse an existing part
-      let itemPart = itemParts[partIndex];
+      itemPart = itemParts[partIndex];
 
       // If no existing part, create a new one
       if (itemPart === undefined) {
-        // If we're creating the first item part, it's startNode should be the
-        // container's startNode
-        let itemStart = this.startNode;
-
-        // If we're not creating the first part, create a new separator marker
-        // node, and fix up the previous part's endNode to point to it
-        if (partIndex > 0) {
-          const previousPart = itemParts[partIndex - 1];
-          itemStart = previousPart.endNode = document.createTextNode('');
-          this._insert(itemStart);
-        }
-        itemPart = new NodePart(itemStart, this.endNode, this.templateFactory);
+        itemPart = new NodePart(this.templateFactory);
         itemParts.push(itemPart);
+        if (partIndex === 0) {
+          itemPart.appendIntoPart(this);
+        } else {
+          itemPart.insertAfterPart(itemParts[partIndex - 1]);
+        }
       }
       itemPart.setValue(item);
       partIndex++;
     }
 
-    if (partIndex === 0) {
-      this.clear();
-      this._value = undefined;
-    } else if (partIndex < itemParts.length) {
-      const lastPart = itemParts[partIndex - 1];
-      // Truncate the parts array so _previousValue reflects the current state
+    if (partIndex < itemParts.length) {
```

Comment:
Can you explain why the other part of the conditional went away?

Reply:
I just realized that I could inline the conditional into the call to `clear()`